### PR TITLE
Migrate min(dim=?) from THC to ATen

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Related issue: https://github.com/pytorch/pytorch/issues/36900

Since I feel this PR is already large enough, I didn't migrate `max` in this PR. Legacy code is not cleaned up either. All these remaining work will be done in later PRs after this is merged.

Benchmark on an extreme case

```python
import torch
print(torch.__version__)

t = torch.randn(100000, 2, device='cuda')

warmup = torch.arange(100000000)
torch.cuda.synchronize()

%timeit t.min(dim=0); torch.cuda.synchronize()
```
Before: 4ms; After: 24.5us.

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -38,154 +38,6 @@ DEFINE_DISPATCH(argmin_stub);
 DEFINE_DISPATCH(cumsum_stub);
 DEFINE_DISPATCH(cumprod_stub);
 
-#define OPTION_TYPE_EQUALITY_CHECK(option, out, self) \
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/ReduceOps.cpp:41]: All these are moved to `aten/src/ATen/native/ReduceOpsUtils.h` for visibility in CUDA.

**Comment by zasdfgbnm** [comment]
close this in favor of #38440
